sopel/modules/ip.py

```diff
@@ -10,52 +10,242 @@
+    "192.184.10.118",
+    ]
+
+IRCCLOUD_REASON = "IRCCloud"
+
+MIBBIT_IP = [
+    "207.192.75.252",
+    "64.62.228.82",
+    "78.129.202.38",
+    "109.169.29.95"
+    ]
+
+MIBBIT_REASON = "Mibbit"
+
+#Hardcoded for safety
+EXEMPT_IP = [
+    ("104.248.43.234", "Lazar, DigitalOcean"),
+    ("13.59.180.136", "War_ Limnoria bot"),
+    ("138.68.23.34", "Approved bot (Idlebot, aismallard)"),
+    ("18.132.171.104", "TARS"),
+    ("3.136.223.150", "bluesoul"),
+    ("54.174.11.206", "Helen"),
+    ("69.115.75.7", "Kufat and bots"),
+    ("94.159.196.226", "docazra, longtime user, IP is on dnsbl"),
+    ("2604:a880:2:d0::250:9001", "Approved bot (Idlebot, aismallard)"),
+    ("91.132.86.177", "bluesoul's bouncer"),
+    ("2001:470:1f07:13b::/64", "kufat's tunnelbroker"),
+    ]
 
 LOGGER = logging.getLogger(__name__)
-
+who_reqs = {}  # Keeps track of reqs coming from this plugin, rather than others
+
+# This dict will mostly be used as a list (walking each element) but having it as a dict
+# is useful for preventing duplicate entries.
+exemptions = {ipaddress.ip_network(i):reason for i, reason in EXEMPT_IP}
+exemptions.update(( (ipaddress.ip_network(i), IRCCLOUD_REASON) for i in IRCCLOUD_IP))
+exemptions.update(( (ipaddress.ip_network(i), MIBBIT_REASON) for i in MIBBIT_IP))
+
+BASE = declarative_base()
+
+safe_mode = True
+db_populated = False
+ipqs_lock = threading.Lock()
+
+# This table will only receive inserts, not updates.
+class KnownIPs(BASE):
+    __tablename__ = 'known_ips'
+    ip = Column(String(50), primary_key=True, unique=True, index=True)
+    score = Column(Float)
+    flag_recent_abuse = Column(Boolean)
+    flag_is_proxy = Column(Boolean)
+    insert_time = Column(TIMESTAMP, server_default=sqlalchemy.sql.func.now())
+
+class ExemptIPs(BASE):
+    __tablename__ = 'exempt_ips'
+    ip = Column(String(50), primary_key=True, unique=True, index=True)
+    # Define type of exemption.
+    # GLine username e.g. uid123@* if true, no g-line at all if false. For IRCCloud.
+    gline_username = Column(Boolean)
+    exempt_reason = Column(Text)
+
+# Existing rows in this table will be updated when a user is seen
+# from the same nick/IP combination.
+# TODO Deferred feature.
+class KnownUsers(BASE):
+    __tablename__ = 'known_users'
+    nick = Column(String(40), primary_key=True)
+    ip = Column(String(50), primary_key=True)
+    cloaked_host = Column(String(40))
+    last_seen = Column(TIMESTAMP, server_default=sqlalchemy.sql.func.now())
 
 class GeoipSection(StaticSection):
     GeoIP_db_path = FilenameAttribute('GeoIP_db_path', directory=True)
     """Path of the directory containing the GeoIP database files."""
-
+    IPQS_key = ValidatedAttribute('IPQS_key')
+    warn_threshold = ValidatedAttribute('warn_threshold', parse=float, default=50.0)
+    malicious_threshold = ValidatedAttribute('malicious_threshold', parse=float, default=70.0)
+    warn_chans = ListAttribute("warn_chans")
+    protect_chans = ListAttribute("protect_chans")
```

Comment:
Maybe it would make sense to move database stuff into its own module?

Reply:
Okay. Though would a directory with an `__init__.py` work, if you wanted to split into files but still have one module?